pyccel/parser/semantic.py

```diff
@@ -3785,212 +3911,263 @@ def _visit_FunctionDef(self, expr):
-                            cls_base = arg_dict[base].cls_base
-                            cls_base.scope.insert_symbol(DottedName(*s.name[1:]))
-
-                for a in arguments:
-                    a_var = a.var
-                    if isinstance(a_var, FunctionAddress):
-                        self.insert_function(a_var)
-                    else:
-                        self.scope.insert_variable(a_var, expr.scope.get_python_name(a.name))
 
-                results = expr.results
-                if results and results[0].annotation:
-                    results = [self._visit(r) for r in expr.results]
-
-                # insert the FunctionDef into the scope
-                # to handle the case of a recursive function
-                # TODO improve in the case of an interface
-                recursive_func_obj = FunctionDef(name, arguments, results, [])
-                self.insert_function(recursive_func_obj)
-
-                # Create a new list that store local variables for each FunctionDef to handle nested functions
-                self._allocs.append(set())
-                self._pointer_targets.append({})
+            if function_call is not None:
+                is_compatible = self._check_argument_compatibility(function_call, arguments, expr, is_elemental, error=False)
+                if not is_compatible:
+                    self.exit_function_scope()
+                    # remove the new created scope and the function name
+                    self.scope.sons_scopes.pop(name)
+                    if is_interface:
+                        self.scope.remove_symbol(name)
+                    continue
+                found_func = True
+            for a in arguments:
+                a_var = a.var
+                if isinstance(a_var, FunctionAddress):
+                    self.insert_function(a_var)
+                else:
+                    self.scope.insert_variable(a_var, expr.scope.get_python_name(a.name))
+
+            if len(arguments)>0 and arguments[0].bound_argument:
+                if arguments[0].var.cls_base.name != cls_name:
+                    errors.report('Class method self argument does not have the expected type',
+                            severity='error', symbol=arguments[0])
+                for s in expr.scope.dotted_symbols:
+                    base = s.name[0]
+                    if base in arg_dict:
+                        cls_base = arg_dict[base].cls_base
+                        cls_base.scope.insert_symbol(DottedName(*s.name[1:]))
+
+            results = expr.results
+            if results and results[0].annotation:
+                results = [self._visit(r) for r in expr.results]
+
+            # insert the FunctionDef into the scope
+            # to handle the case of a recursive function
+            # TODO improve in the case of an interface
+            recursive_func_obj = FunctionDef(name, arguments, results, [])
+            self.insert_function(recursive_func_obj)
```

Comment:
Ok. Then this is about #551

Reply:
I am curious if your implementation fixes this problem. It seems to me that it probably could in some cases

Reply:
I'm not even sure how we can implement it, imagine this case
```python
def f(x:'int|float'):
    if abs(x) < 1:
        print(x)
    elif x > 1:
        f(-int(x)+1)
    else:
        f(-x/2)
```
this function is recursive in Python, but is it recursive in Fortran?

Reply:
In Fortran we don't have the notion of a recursive Interface, so it is harder to implement

Reply:
I don't know if this is possible in Fortran but I think your PR would fix a case like:
```python
def f(x:'int|float'):
    if abs(x) < 1:
        print(x)
    elif x > 1:
        f(-int(x)+1)
    else:
        f(-int(x/2))
```

Reply:
I agree. In fact in #551 we suggested raising an error for all cases of non-trivial interface recursivity. Your PR would fix some of the less pathological cases though. Such as the ones mentioned in the issue. We would then just be missing an error for the unmanageable (in Fortran) cases like the one you suggest. But I think this can be left for a separate PR